wavefront/resource_alert.go

```diff
@@ -344,3 +366,42 @@ func validateThresholdLevels(m map[string]string) error {
 	}
 	return nil
 }
+
+// Decodes the runbook links from the state and returns a []string of links
+func decodeRunbookLinks(d *schema.ResourceData) (links []string) {
+	for _, link := range d.Get(runbookLinksKey).([]interface{}) {
+		links = append(links, link.(string))
+	}
+	return links
+}
+
+func decodeAlertTriageDashboards(d *schema.ResourceData) []wavefront.AlertTriageDashboard {
+	alertTriageDashboards := []wavefront.AlertTriageDashboard{}
+
+	if dashboards, ok := d.Get(alertTriageDashboardsKey).([]interface{}); ok {
+		for _, dashboard := range dashboards {
+			dashboardData := dashboard.(map[string]interface{})
+			alertTriageDashboard := wavefront.AlertTriageDashboard{
+				DashboardId: dashboardData[dashboardIDKey].(string),
+				Description: dashboardData[descriptionKey].(string),
+				Parameters:  make(map[string]map[string]string),
+			}
+
+			if parameters, ok := dashboardData[parametersKey].([]interface{}); ok && len(parameters) > 0 {
+				// Assuming there should be only one parameters block
+				parametersData := parameters[0].(map[string]interface{})
+
+				if constants, ok := parametersData[constantsKey].(map[string]interface{}); ok {
+					alertTriageDashboard.Parameters[constantsKey] = make(map[string]string)
+					for key, value := range constants {
+						alertTriageDashboard.Parameters[constantsKey][key] = value.(string)
+					}
+				}
+			}
+
+			alertTriageDashboards = append(alertTriageDashboards, alertTriageDashboard)
+		}
+	}
+
+	return alertTriageDashboards
+}
```

Comment:
This method is fairly long and has a fair amount of nesting which adds a fair amount of complexity and probably requires a more complex test to do it justice. (Plus other cyclomatic complexity side effects.)

You might consider breaking it up into multiple functions. Naively I would try pulling the dashboard parsing out. Maybe the param parsing, but that could be excessive. In theory it should make unit tests simpler and make the functions easier to understand. (Sometimes I also find this process sheds light on a simplification I didn't notice before the refactor.)

Not a blocker though, the function is simple enough that I can follow it. Just a smell that I see at first glance.

Reply:
Yeah agreed! I'll take a look at simplifying this too today.